src/widgets/ansi_string_buffer.rs

```diff
@@ -0,0 +1,171 @@
+use std::fmt::{self, Display, Formatter};
+
+use crate::prelude::*;
+
+/// A buffer that allows widgets to easily be rendered to a string with ANSI escape sequences.
+pub struct AnsiStringBuffer {
+    pub area: Rect,
+    buf: Buffer,
+}
+
+impl AnsiStringBuffer {
+    pub fn new(width: u16, height: u16) -> Self {
+        Self {
+            area: Rect::new(0, 0, width, height),
+            buf: Buffer::empty(Rect::new(0, 0, width, height)),
+        }
+    }
+
+    pub fn render_ref(&mut self, widget: &impl WidgetRef) {
+        widget.render_ref(self.area, &mut self.buf);
+    }
+}
+
+impl Display for AnsiStringBuffer {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let mut last_style = None;
+        for y in 0..self.area.height {
+            if y > 0 {
+                f.write_str("\n")?;
+            }
+            for x in 0..self.area.width {
+                let cell = self.buf.get(x, y);
+                let style = (cell.fg, cell.bg, cell.modifier);
+                if last_style.is_none() || last_style != Some(style) {
+                    write_cell_style(cell, f)?;
+                    last_style = Some(style);
+                }
+                f.write_str(cell.symbol())?;
+            }
+        }
+        f.write_str("\u{1b}[0m")
+    }
+}
+
+fn write_cell_style(cell: &buffer::Cell, f: &mut Formatter<'_>) -> fmt::Result {
```

Comment:
it _feels_ more natural to me to have the order of these arguments reversed:

```suggestion
fn write_cell_style(f: &mut Formatter<'_>, cell: &buffer::Cell) -> fmt::Result {
```

Thoughts?

Reply:
Agreed!. Done